### PR TITLE
Add read functionality for InSituVane tests

### PR DESCRIPTION
**Pull request by jtrainor1**
### Issues addressed by this PR
https://github.com/BHoM/AGS_Toolkit/issues/33
Addresses issue 33, adding read and convert capabilities for In Situ Vane data from AGS files.

See also
https://github.com/BHoM/AGS_Toolkit/issues/26
https://github.com/BHoM/BHoM/issues/1640
https://github.com/BHoM/BHoM_Engine/issues/3418

### Test files
Grasshopper:
https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/AGS_Toolkit/%2333-addinsituvanereadtests/37367%20-%202023-08-08%201339%20-%20Final%20-%202.ags?csf=1&web=1&e=IiVlCW

### Changelog
- Added bug fix for date formats that are yyyy-mm-dd as C# would parse the mm as minutes instead of months;
- Added real functionality for `InSituVaneTests` and all associated properties.

### CODE REVIEW

File: AGS_Adapter/Convert/Units/Units.cs [hunk 1]
```diff
@@ -106,6 +106,11 @@ public static double Units(double value, string unit, string key = "")
                     return value.FromSiemensPerCentimetre()*Math.Pow(10,-6);
                 case "ms/cm":
                     return value.FromSiemensPerCentimetre() * Math.Pow(10, -3); ;
+                // Pressure
+                case "kPa":
+                case "kpa":
+                case "KPA":
```
Comment (peterjamesnugent) [inline AGS_Adapter/Convert/Units/Units.cs]:
```suggestion
                case "kpa":
```
Only lowercase version required as all white space is removed and `.ToLower()` is used.

File: AGS_Adapter/CRUD/Read/InSituVane.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+namespace BH.Adapter.AGS
+{
+    public partial class AGSAdapter
+    {
+        /***************************************************/
+        /**** Private Methods                           ****/
+        /***************************************************/
+
+        private List<InSituVane> ReadInSituVane(List<string> ids = null)
```
Comment (peterjamesnugent) [inline AGS_Adapter/CRUD/Read/InSituVane.cs]:
```suggestion
        private List<InSituVane> ReadInSituVanes(List<string> ids = null)
```
Read are plural because you're reading all, Convert is singular as you are converting a single item.

File: AGS_Adapter/Convert/FromAGS/InSituVane.cs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+            // InSituVane Reference
+            string details = GetString(data, "IVAN_REM");
+            string weather = GetString(data, "IVAN_ENV");
+            DateTime testdate = GetDateTime(data, units, "IVAN_DATE");
+            string stratumreference = GetString(data, "GEOL_STAT");
+            string files = GetString(data, "FILE_FSET");
+
+            InSituVaneReferenceProperties InSituVaneReferenceProperties = new InSituVaneReferenceProperties() { Details = details, Weather = weather, Date = testdate, StratumReference = stratumreference, FileReference = files };
+            if (InSituVaneReferenceProperties != null)
+                testProperties.Add(InSituVaneReferenceProperties);
```
Comment (peterjamesnugent) [inline AGS_Adapter/Convert/FromAGS/InSituVane.cs]:
```suggestion
            InSituVaneReferenceProperties InSituVaneReferenceProperties = new InSituVaneReferenceProperties() 
            { 
                Details = details, 
                Weather = weather, Date = testdate, 
                StratumReference = stratumreference, 
                FileReference = files 
            };

            if (InSituVaneReferenceProperties != null)
                testProperties.Add(InSituVaneReferenceProperties);
```
Formatting to help readability

File: AGS_Adapter/Convert/FromAGS/InSituVane.cs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            double result = GetDouble(data, units, "IVAN_IVAN");
+
+            List<ITestProperties> testProperties = new List<ITestProperties>();
+
+            // InSituVane Reference
+            string details = GetString(data, "IVAN_REM");
+            string weather = GetString(data, "IVAN_ENV");
+            DateTime testdate = GetDateTime(data, units, "IVAN_DATE");
+            string stratumreference = GetString(data, "GEOL_STAT");
+            string files = GetString(data, "FILE_FSET");
```
Comment (peterjamesnugent) [inline AGS_Adapter/Convert/FromAGS/InSituVane.cs]:
```suggestion
            string details = GetString(data, "IVAN_REM");
            string weather = GetString(data, "IVAN_ENV");
            DateTime testDate = GetDateTime(data, units, "IVAN_DATE");
            string stratumReference = GetString(data, "GEOL_STAT");
            string files = GetString(data, "FILE_FSET");
```
Camel case is used for variable names.

File: AGS_Adapter/Convert/FromAGS/InSituVane.cs [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+            if (InSituVaneReferenceProperties != null)
+                testProperties.Add(InSituVaneReferenceProperties);
+
+            // InSituVane Test Properties
+            string testReference = GetString(data, "IVAN_TESN");
+            string type = GetString(data, "IVAN_TYPE)");
+            string tester = GetString(data, "IVAN_CONT"); 
+            string testmethod = GetString(data, "IVAN_METH");
+            string accreditingbody = GetString(data, "IVAN_CRED");
+            string testStatus = GetString(data, "TEST_STAT");
```
Comment (peterjamesnugent) [inline AGS_Adapter/Convert/FromAGS/InSituVane.cs]:
```suggestion
            string testReference = GetString(data, "IVAN_TESN");
            string type = GetString(data, "IVAN_TYPE)");
            string tester = GetString(data, "IVAN_CONT"); 
            string testMethod = GetString(data, "IVAN_METH");
            string accreditingBody = GetString(data, "IVAN_CRED");
            string testStatus = GetString(data, "TEST_STAT");
```

File: AGS_Adapter/Convert/FromAGS/InSituVane.cs [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+            InSituVaneTestProperties InSituVaneTestProperties = new InSituVaneTestProperties()
+            {
+                Reference = testReference,
+                Type = type,
+                Method = testmethod,
+                Tester = tester,
+                AccreditingBody = accreditingbody,
+                Status = testStatus
+            };
+            if (InSituVaneTestProperties != null)
```
Comment (peterjamesnugent) [inline AGS_Adapter/Convert/FromAGS/InSituVane.cs]:
```suggestion
            };
            
            if (InSituVaneTestProperties != null)
```

File: AGS_Adapter/Convert/FromAGS/InSituVane.cs [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+                testProperties.Add(InSituVaneTestProperties);
+
+            // InSituVane Result Properties
+            double vaneresidualresult = GetDouble(data, units, "IVAN_IVAR");
```
Comment (peterjamesnugent) [inline AGS_Adapter/Convert/FromAGS/InSituVane.cs]:
```suggestion
            double vaneResidualResult = GetDouble(data, units, "IVAN_IVAR");
```

File: AGS_Adapter/Convert/FromAGS/InSituVane.cs [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+
+            InSituVaneResultProperties InSituVaneResultProperties = new InSituVaneResultProperties()
```
Comment (peterjamesnugent) [inline AGS_Adapter/Convert/FromAGS/InSituVane.cs]:
```suggestion
            InSituVaneResultProperties inSituVaneResultProperties = new InSituVaneResultProperties()
```

File: AGS_Adapter/Convert/FromAGS/InSituVane.cs [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+            string stratumreference = GetString(data, "GEOL_STAT");
+            string files = GetString(data, "FILE_FSET");
+
+            InSituVaneReferenceProperties InSituVaneReferenceProperties = new InSituVaneReferenceProperties() { Details = details, Weather = weather, Date = testdate, StratumReference = stratumreference, FileReference = files };
+            if (InSituVaneReferenceProperties != null)
+                testProperties.Add(InSituVaneReferenceProperties);
+
+            // InSituVane Test Properties
+            string testReference = GetString(data, "IVAN_TESN");
+            string type = GetString(data, "IVAN_TYPE)");
```
Comment (peterjamesnugent) [inline AGS_Adapter/Convert/FromAGS/InSituVane.cs]:
```suggestion
            string type = GetString(data, "IVAN_TYPE");
```
This comes up in the warning for the Pull, should have been caught.

**Review by peterjamesnugent — CHANGES_REQUESTED** [review]
I have reviewed and updated the test script to be more thorough - I suggest you test against a few AGS files as quite a few fields are empty.

I suggest you use the script for a template when you add in the other tests.

I also found a bug with the date conversion which I have updated.

**Review by peterjamesnugent — APPROVED** [review]
Happy this works as intended now with the commited changes.

**Comment by peterjamesnugent** [comment]
@BHoMBot check required

**Comment by peterjamesnugent** [comment]
@BHoMBot check copyright-compliance

**Comment by peterjamesnugent** [comment]
@BHoMBot check ready-to-merge
